acme_diags/driver/qbo_driver.py

```diff
@@ -0,0 +1,178 @@
+        ampN[ilev] = np.sqrt(2 * pspecN[ilev]) * (f[1] - f[0])
+    return pspecN, ampN
+
+
+def process_u_for_power_spectral_density(data_region):
+    # Average over vertical levels and horizontal area
+    levBot = 22
+    levTop = 18
+    # Average over lat and lon
+    data_lat_lon_average = cdutil.averager(data_region, axis='xy')
+    # Average over vertical
+    x0 = np.nanmean(np.array(data_lat_lon_average(level=(levTop, levBot))), axis=1)
+    # x0 should now be 1D
+    return x0
+
+
+def ceil_log2(x):
+    """
+    Given a number, calculate the exponent for the next power of 2.
+
+    Example:
+        ceil_log2(16) = 4
+        ceil_log2(17) = 5
+    """
+    return np.ceil(np.log2(x)).astype('int')
+
+
+def get_psd_from_deseason(xraw, periodNew):
+    x0 = deseason(xraw)
+
+    # Sampling frequency: assumes frequency of sampling = 1 month
+    Fs0 = 1
+    T0 = 1 / Fs0
+    L0 = len(xraw)
+    t0 = np.arange(0, L0) * T0
+    NFFT0 = 2 ** ceil_log2(L0)
+
+    # Apply fft on x0 with n = NFFT
+    X0 = scipy.fftpack.fft(x0, n=NFFT0) / L0
+    # Frequency (cycles/month)
+    f0 = Fs0 * np.arange(0, (NFFT0 / 2 + 1)) / NFFT0
+
+    period0 = 1 / f0
+
+    # Amplitude
+    AMP0 = 2 * abs(X0[0:int(NFFT0 / 2 + 1)])
+    # Power spectral density
+    PSDx0 = AMP0 ** 2 / L0
+    Tend0 = T0 * L0
+    # Total spectral power
+    Pxf0 = Tend0 * np.sum(PSDx0)
+    # TODO: Because why?
+    # Need to flip original period, AMP, and PSDx0 because
+    period0_flipped = period0[::-1]
+    AMP0_flipped = AMP0[::-1]
+    PSDx0_flipped = PSDx0[::-1]
+
+    AMPnew0 = np.interp(periodNew, period0_flipped[:-1], AMP0_flipped[:-1])
+    PSDxnew0 = np.interp(periodNew, period0_flipped[:-1], PSDx0_flipped[:-1])
+    return PSDxnew0, AMPnew0
+
+
+def run_diag(parameter):
+    variables = parameter.variables
+    # The region will always be 5S5N
+    region = '5S5N'
+    test_data = utils.dataset.Dataset(parameter, test=True)
+    ref_data = utils.dataset.Dataset(parameter, ref=True)
+    for variable in variables:
+        if parameter.print_statements:
+            print('Variable={}'.format(variable))
+        test_var = test_data.get_timeseries_variable(variable)
+        ref_var = ref_data.get_timeseries_variable(variable)
+        qbo_region = default_regions.regions_specs[region]['domain']
+
+        test_region = test_var(qbo_region)
+        ref_region = ref_var(qbo_region)
+
+        test = {}
+        ref = {}
+
+        # Diagnostic 1: average over longitude & latitude to produce time,height array of u field:
+        # https://en.wikipedia.org/wiki/Quasi-biennial_oscillation
+        # U = "monthly-mean, zonal-mean equatorial zonal wind"
+        test['qbo'], test['level'] = process_u_for_time_height(test_region)
+        ref['qbo'], ref['level'] = process_u_for_time_height(ref_region)
+
+        # Diagnostic 2: calculate and plot the amplitude of wind variations with a 20-40 month period
+        # TODO: What is pspec?
```

Comment:
What is `pspec`?

Reply:
pspec is double the sum of the power spectral density between 20 and 40 days and exists mainly as an intermediary variables to obtain amp.